### ISSUE TITLE
Handling data dependencies and reproducability 

### ISSUE THREAD

**Issue by AlexRobson** [ScottishCovidResponse/SCRCIssueTracking#229]
Within `Simulation.jl` there are various data dependencies that may change as different versions of these data appear. 

There are however probably two uses cases:

* One scenario is the 'get me the exame same version always'). DataDeps can sort out getting a file dependency onto a local set-up (such as from a website). Checksums are used to validate the file. In this scenario, datadeps is goof for static datasets that are versioned outside of julia (such as the data provider).

The opposite scenario may be that is the 'get me the latest data always'. This is different and may require some thought on how to version this (other than labelling the time stamp of that data and making it a static dataset in the sense of data deps, which falls back to DataDeps. A possible package for this (referenced from datadeps itself) is `RemoteFiles.jl` (I have not used this package before). 

This is an example of the `DataDeps` set up. This is a snippet from the WIP in https://github.com/ScottishCovidResponse/SCRCIssueTracking/issues/192

```
using Simulation
using Unitful
using Unitful.DefaultSymbols
using Simulation.Units
using DataDeps
import Pkg


# Data used for Testing loading of parameters
# https://github.com/ScottishCovidResponse/SCRCIssueTracking/issues/192
# DataDeps is definitely overkill for this, but this set up
# can be used for the larger data files.
register(
    DataDep(
        "TestFile-Parameters",
        "Param.toml has two parameters for beta_env and beta force",
        "https://www.gla.ac.uk/media/Media_722505_smxx.toml",
        "5cd42a53e0ef51c10c5d5efe20df2d52194e71400e232294aa53d3053765599e",
        fetch_method = DataDeps.fetch_http
    )
)

p = Pkg.TOML.parsefile(datadep"TestFile-Parameters"))
```

**Comment by sdl1** [comment]
It could be good to have all this stuff defined in a separate lightweight package, both for ease of maintenance and for re-usability across different julia projects.

**Comment by AlexRobson** [comment]
^ @sdl1 I think I may know what you mean, but could you sketch out more details? Something like a (for sake of argument, because naming things is hard), a `EpiDataZoo` ?

**Comment by AlexRobson** [comment]
```
using Simulation
using EpiDataZoo <--- handles DataDeps registries

scotpop = load(datadeps"EpiDataZoo.ScotPopData")
```

**Comment by richardreeve** [comment]
Yes, I'm a little bit unclear. If a lightweight package makes sense then great, but can we do something automated like [DataDepsGenerators.jl](https://github.com/oxinabox/DataDepsGenerators.jl) and some continuous integration when the data repository updates... or is that the kind of thing you mean in fact?

**Comment by sdl1** [comment]
Yes I was just imagining something to store the DataDeps registration blocks / handle updating datadeps / provide a nice interface, as in Alex's example. The advantage of a separate package would be decoupling Simulation releases from data releases, and allowing other Julia models to use it without needing to have Simulation.jl as a dependency. But I don't think having a separate package is a priority, we could have it within Simulation for now.

**Comment by richardreeve** [comment]
I'm not concerned about it being a separate package tbh, although them not being in the registry will make installation more complicated I guess. However, looking at the DataDeps pages, can't we have something that automatically handles any of our datasets - the idea is that all of the relevant information to access a given dataset will be both (a) in a human readable form (toml) in one of our git repositories along with the most important associated metadata and (b) in a Postgres database including all of the metadata. So everything will be pretty standardised (two standards I think, one for single parameters, one for actual datasets).

Doesn't that mean that we can have an automated system to access any datadep related to our project?

Maybe I confused things by providing a couple of files to download that are pretty arbitrary - if so, sorry - they were really just for the format...

**Comment by richardreeve** [comment]
@sdl1 @AlexRobson I've been thinking about this... maybe this is exactly what the separate package is that you're thinking of, but can we develop something that is a wrapper around `DataDeps`? So it uses `scrc"human/demographics/scotland"` to access the `human/demographics/scotland` metadata, cache the the most recent version of the dataset? Or just `scrc"human/demographics/scotland:public@v1.0"` to access the specific version of the dataset directly. Either way, a second argument to the read function would allow us to get hold of the `groupages-10k-grid` object within that. Because it's not `DataDeps` directly, it can fudge the issue of which version it's accessing or not according to need... does that make sense / seem like a good idea?

**Comment by AlexRobson** [comment]
I've started to flesh this out a bit under a new (WIP/prototype) package here: 
https://github.com/ScottishCovidResponse/SimulationData

I don't think we need any special wrapping around `DataDeps` to handle version control. We can just create a DataDep for `ScotPop-v1`, `ScotPop-v2`, etc. 

We can probably include some mechanism to download the latest (unversioned) data. 

Just a thought while I was working on this, we have a Simulation dependency entirely because of the Unitful units in validation. May be worht considering seperating those out into a separate package too , i.e. an `EpidemiologyUnits` or somesuch. 

It's irrelevant of course if we just want this to 'get the data, don't do validation' only. For the two test files I had anyway, I added in some validation.

**Comment by richardreeve** [comment]
Thanks @AlexRobson. I keep on meaning to find time to look at this, but in the meanwhile, can I point out this - https://zulip.scrc.uk/#narrow/stream/14-Data/topic/Data.20pipeline.20API/near/5516 - that Chris has written up on how he sees the API working in case you haven't seen it.

**Comment by AlexRobson** [comment]
I think that this isn't really an issue to include. Presumably, then, we want to add this API to simulation (the blocks can stay in SimulationData)

```
read_estimate(data_product, representation) -> estimate
read_distribution(data_product, representation) -> distribution
read_array(data_product, representation) -> array
read_table(data_product, representation) -> table
```

Then, given the way the API has been mocked:

Using ScotPop as an example:
`data_product = "human/demographics/scotland"`
`representation = "grid10k_total"`

In Simulation we do:

`read_estimate(data_product, representation)`

This will have a function something like:
```
using SimulationData

function read_estimate(data_product::AbstractString, representation::AbstractString)
    deps = datadep"data_product*representation" <- This will evaluate to a registered block
    fpath = joinpath(deps, representation*".toml") <- This assumes that representation is also the name and we just add in the file type
   data = h5open(fpath, "r") do file
       read(file, representation) <-- There may be some assumptions on how the h5 is set up
  end
  return data
end
```

**Comment by richardreeve** [comment]
Actually, the _representation_ (though Chris seems to be quietly changing this to _quantity_) is going to be inside the file, so I don't think that it needs to be inside the datadep string. However, the version will be, and there will also be a root for namespace, so we'll end up with something like
```julia
deps = datadep"root:data_product@version"
```
and that will just be the file path.
